### ISSUE TITLE
Global find

### ISSUE THREAD

**Issue by akshayl** [nteract/archived-desktop-app#1817]
I can't seem to figure out a way to find across the entire notebook, Cmd+F only finds in the current cell. Am I missing something or is this a missing feature?

**Comment by rgbkrk** [comment]
Missing feature, thanks for adding it to the tracker.

**Comment by danielgwilson** [comment]
Would love to pile onto this—don't have the chops to implement it myself but definitely a desired feature.

**Comment by gdbassett** [comment]
I would agree on this as a priority.  I am going to have to install a separate notebook server just to view notebooks because I can't search for specific insights in my notebooks. (OSX 0.7.1).

**Comment by bradley-evans** [comment]
I found `nteract` via the `choco` package manager and was really disappointed by the absence of a search and replace function. That's a pretty critically basic IDE feature.

**Comment by montaguegabe** [comment]
+1

**Comment by rgbkrk** [comment]
Since there are multiple issues for this same feature request, I'm going to consolidate them in to this one.

**Comment by rgbkrk** [comment]
Pinning this issue as it should never go stale. We definitely need to implement this one. As our data structures have gotten more normalized, I'm feeling better about going after this.

**Comment by matanox** [comment]
I think if 'advanced search' is implemented, would be nice to be able to choose between some options, crudely separated like below, but other kinds of flexibility might be nicer:

- search selected cells v.s. entire sheet
- search code v.s. results v.s. both

Or something along those lines, in terms of flexibility

**Comment by zhanwenchen** [comment]
This issue is more than 1 year old.

**Comment by jdetle** [comment]
Global search is hard for this kind of thing. I'm afraid I'm a little out of my element on this, but I'm going to go ahead and try to propose a solution.

We should only be searching input text for now, outputs can be left for later work. I think Input Search should be implemented first at the cell level. When we get this working right at the cell level, implementing at the notebook level should be able to be implemented as a map over all the cells with a newly passed `FindText` prop. The work on the notebook level would be around the presentational aspects of switching cell to cell.

The pros of this approach seem to me to follow here:
- The mechanism of switching the viewport to go view another cell is out of scope in this approach. We can focus on the core UI/UX, which is switching through highlighted text inside 
- Encapsulation. `FindText` is a prop, push is always better than pull. We could do something where we react to changes in text across all cells and maintain some in-memory copy of all cell input text, but that seems like we would have to create extra bookkeeping and overhead around that.

A potential bad outcome I could see happening:
- The work on Input Search in the cell gets finished released while nothing gets done on the notebook level find feature, so we are left with some half-baked search functionality that is almost worse than having none at all.

What do people think?

**Comment by jmon12** [comment]
What is the status on this? 
This feature is missing drastically, I actually have to use another editor in parallel just to search some pattern. 
One of the key point of nteract is to have a notebook editor really integrated with the desktop. When you have to read a lot of notebooks, it becomes very handy, but the lack of search functionality makes it of little use!
I'm sorry not to be able to contribute myself, and hope some more qualified people think about implementing it!

**Comment by m1sta** [comment]
> What do people think?

I think every day that there isn't even a basic find feature is another day I rethink whether this tool is for me.

**Comment by Antzee1** [comment]
I love the nteract notebook, but I too think it's hard to use without a search / find option. Also my boyfriend just tried nteract and commented the same thing. Apologies for not being able to contribute.

**Comment by dpsanders** [comment]
Just installed Nteract and almost the first thing I did was to try to search for a particular function in somebody else's notebook. This seems like a high priority to implement.

**Comment by captainsafia** [comment]
Thanks for sharing your feedback, everyone!

This feature is tricky because it'll require some design and technical work to support searching through all the content in the cell efficient, displaying it in the UI, and providing an interface for entering search requests.

If anyone has any thoughts on how to implement this effectively or recommendations for React components/JavaScript packages that might ease the implement work, please let us know.

**Comment by motin** [comment]
Maybe https://github.com/rhysd/electron-in-page-search?

**Comment by captainsafia** [comment]
Looks promising!

Thanks for sharing @motin.

**Comment by hsbarbosa** [comment]
In the meantime, an ugly workaround is to use the Developer tools (View ->Toggle Developer Tools) to search from there. You can right-click on the div element and click on "scroll into view". I know it's a horrible workaround but might be useful for someone else.

**Comment by JMBurley** [comment]
Wanted to check in on feature progress here.  I see that this was originally pinned, but is no longer pinned on the issue board?

Is there an ETA on implementation after @motin 's suggestion?

**Comment by captainsafia** [comment]
@JMBurley No updates at the moment. As nteract is a volunteer-run project, it's difficult to prioritize every feature and this one has fallen into the backlong in favor of bug fixes. I'll go ahead and tag this as help wanted in case anyone wants to contribute in this effort.

**Comment by jdetle** [comment]
It looks like the electron in page search doesn't have a 'replace' feature. We definitely won't have the ability to search with a native implementation as we implement virtualization.

Here's what I'm thinking:
1. Implement the electron search and see if it can work. It would certainly be better than nothing. 
2. Implement search and replace for code cells and markdown cells. CodeMirror has functionality for this already. For the markdown cells we can try to imitate the codemirror ui/ux as closely as possible.
3. If step 2 works, we can implement global search and replace by mapping across all cells, controlling that search and replace feature.

Does that make sense? What do people think?

**Comment by captainsafia** [comment]
That sounds like a good plan. How hard is #2 to implement?

**Comment by jdetle** [comment]
I'm going to investigate to see how hard it is to hook up. As you know, codemirror is a fickle beast.

**Comment by samialsindi** [comment]
Adding my voice in to the mix, I love this app but I can't move to it until a basic global find is incorporated

**Comment by lwbaqueros** [comment]
+1

**Comment by andreimargeloiu** [comment]
I thought it's impossible not to be a global find until seeing this thread...

**Comment by knkumar** [comment]
Search of any form would be appreciated, cell, global etc

**Comment by deeplook** [comment]
Seems like it's going to be soon four year to find a find feature, amazing! ;)

**Comment by lwbaqueros** [comment]
It is our 4th anniversary here 🤣😂😅😀😐😣😥😭
